pkg/liquidity-source/balancer-v2/stable/pool_simulator.go

```diff
@@ -150,17 +232,316 @@ func (s *PoolSimulator) CalcAmountOut(params pool.CalcAmountOutParams) (*pool.Ca
+		return nil, ErrBatchSwapDisabled
+	}
+
+	if indexIn >= 0 && indexOut >= 0 {
+		return s.swapDirect(tokenAmountIn.Token, tokenOut, amountIn)
+	}
+
+	if indexIn < 0 && indexOut >= 0 {
+		return s.swapFromBase2Main(tokenAmountIn.Token, tokenOut, amountIn)
+	}
+
+	if indexIn >= 0 && indexOut < 0 {
+		return s.swapFromMain2Base(tokenAmountIn.Token, tokenOut, amountIn)
+	}
+
+	return s.swapBetweenBasePools(tokenAmountIn.Token, tokenOut, amountIn)
+}
+
+func (s *PoolSimulator) swapDirect(tokenIn, tokenOut string, amountIn *uint256.Int) (*pool.CalcAmountOutResult, error) {
+	amountOut, err := s.OnSwap(tokenIn, tokenOut, amountIn)
+	if err != nil {
+		return nil, err
+	}
+
+	return s.buildSwapResult(tokenOut, amountOut, nil), nil
+}
+
+func (s *PoolSimulator) swapFromBase2Main(tokenIn, tokenOut string, amountIn *uint256.Int) (*pool.CalcAmountOutResult, error) {
+	basePool, err := s.getBasePool(tokenIn)
+	if err != nil {
+		return nil, err
+	}
+
+	bptToken := basePool.GetAddress()
+
+	var (
+		hops = make([]shared.Hop, 0, 2)
+
+		bptAmount *uint256.Int
+		joinIndex *big.Int
+	)
+
+	switch basePool.GetType() {
+	case composablestable.DexType:
+		bptAmount, err = basePool.OnSwap(tokenIn, bptToken, amountIn)
+	default:
+		bptAmount, err = basePool.OnJoin(tokenIn, amountIn)
+		joinIndex = shared.PackJoinExitIndex(shared.PoolJoin, 0)
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	hops = append(hops, shared.Hop{
+		PoolId:        basePool.GetPoolId(),
+		Pool:          basePool.GetAddress(),
+		TokenIn:       tokenIn,
+		TokenOut:      bptToken,
+		AmountIn:      amountIn.ToBig(),
+		AmountOut:     bptAmount.ToBig(),
+		JoinExitIndex: joinIndex,
+	})
+
+	amountOut, err := s.OnSwap(tokenOut, bptToken, bptAmount)
+	if err != nil {
+		return nil, err
+	}
+
+	hops = append(hops, shared.Hop{
+		PoolId:    s.GetPoolId(),
+		Pool:      s.GetAddress(),
+		TokenIn:   bptToken,
+		TokenOut:  tokenOut,
+		AmountIn:  bptAmount.ToBig(),
+		AmountOut: amountOut.ToBig(),
+	})
+
+	return s.buildSwapResult(tokenOut, amountOut, hops), nil
+}
+
+func (s *PoolSimulator) swapFromMain2Base(tokenIn, tokenOut string, amountIn *uint256.Int) (*pool.CalcAmountOutResult, error) {
+	basePool, err := s.getBasePool(tokenOut)
+	if err != nil {
+		return nil, err
+	}
+
+	bptToken := basePool.GetAddress()
+
+	var hops = make([]shared.Hop, 0, 2)
+
+	bptAmount, err := s.OnSwap(tokenIn, bptToken, amountIn)
+	if err != nil {
+		return nil, err
+	}
+
+	hops = append(hops, shared.Hop{
+		PoolId:    s.poolID,
+		Pool:      s.GetAddress(),
+		TokenIn:   tokenIn,
+		TokenOut:  bptToken,
+		AmountIn:  amountIn.ToBig(),
+		AmountOut: bptAmount.ToBig(),
+	})
+
+	var (
+		amountOut *uint256.Int
+		exitIndex *big.Int
+	)
+
+	switch basePool.GetType() {
+	case composablestable.DexType:
+		amountOut, err = basePool.OnSwap(bptToken, tokenOut, bptAmount)
+	default:
+		amountOut, err = basePool.OnExit(tokenOut, bptAmount)
+		exitIndex = shared.PackJoinExitIndex(shared.PoolExit, 1)
```

Comment:
```suggestion
		exitIndex = shared.PackJoinExitIndex(shared.PoolExit, len(hops))
```
Should use len(hops) instead of hardcoded value